### ISSUE TITLE
Error while building kokkos in Trilinos package 

### ISSUE THREAD

**Issue by Resacd** [kokkos/kokkos#5004]
I'm trying to build the Epetra, AzetcOO, Belos,ML and Ifpack packages of Trilinos downloaded here https://github.com/trilinos/Trilinos. I'm using the last version of cmake and the compiler is VS 2022. The problem is after configuring and generating the solution when I try to build ALL_BUILD I'm facing the following compiler error which I don't know how to fix it.
Error C2953 'Kokkos::Impl::FunctorAnalysis<PatternInterface,Policy,Functor>::DeduceTeamShmem<F,
std::enable_if<0<,void>::type>': class template has already been defined kokkoskernels

And the error refers to this part of the KOKKOS,
```
template <class F>
  struct DeduceTeamShmem<
      F, typename std::enable_if<0 < sizeof(&F::team_shmem_size)>::type> {
    enum : bool { value = true };

    static size_t team_shmem_size(F const* const f, int team_size) {
      return f->team_shmem_size(team_size);
    }
  };

  template <class F>
  struct DeduceTeamShmem<
      F, typename std::enable_if<0 < sizeof(&F::shmem_size)>::type> {
    enum : bool { value = true };

    static size_t team_shmem_size(F const* const f, int team_size) {
      return f->shmem_size(team_size);
    }
  };
```

**Comment by dalg24** [comment]
You need to provide more information.  What version of Trilinos are you using.  How did you configure?  Please provide a bit more context with your error message.  The trailing "kokkoskernels" makes little sense (typo?)

**Comment by ajpowelsnl** [comment]
Hi @Resacd - would you please provide a minimal reproducer for this error?  Please include compiler version information, in addition to any other details pertinent to reproducing the error.

**Comment by ajpowelsnl** [comment]
Hello @Resacd -- are you still having problems with these builds?  If so, please let @dalg24 and me know, and we'll do our best to help you out.

**Comment by Resacd** [comment]
I built the Trilinos with kokkos turned off, as a result I was not able to have Tpetra due to its dependency on kokkos. I just don't the reason for this error.

**Comment by Resacd** [comment]
I downloaded the package from github, the latest one. Then, I wrote my one shell script and build the package using mingw. After compiling I got the error that I posted above.

**Comment by ajpowelsnl** [comment]
> I downloaded the package from github, the latest one. Then, I wrote my one shell script and build the package using mingw. After compiling I got the error that I posted above. Another time I followed the https://github.com/INMOST-DEV/INMOST/wiki/0207-Compilation-Trilinos-Windows and the same problem was there.

OK, thank you.  We'll look into it, and get back to you soon with a path forward.

**Comment by dalg24** [comment]
There is still not enough information to help.
> I downloaded the package from github, the latest one.

Did you download a release tarball or did you clone the repository?  This is ambiguous.

Also what options did you set when you configured Trilinos with cmake?

> Error C2953 'Kokkos::Impl::FunctorAnalysis<PatternInterface,Policy,Functor>::DeduceTeamShmem<F,
std::enable_if<0<,void>::type>': class template has already been defined kokkoskernels

Can you share the build log?  What source file was being compiled when you got that error?

**Comment by ajpowelsnl** [comment]
Hello @Resacd -- would you please provide answers to the following:

```
**Describe the bug**
Please provide a concise, clear description of the bug, as well as any available error logs.

**Please also include the following items to support reproducing the bug**
1. compilers (with versions)
2. Kokkos release or commit used (i.e., the sha1 number)
3. platform and backend
4. cmake configure command
5. output from cmake command
6. code needed to reproduce the bug
7. command line needed to reproduce the bug
7. please also attach the `KokkosCore_config.h` header file (generated during the build);
**Any additional info**
Please provide any additional context about the issue here.
```

In a word, we need a step-by-step reproducer to be able to help.

**Comment by ajpowelsnl** [comment]
Hi @Resacd -- Are you still having Kokkos-related build issues with the current version of Trilinos?  If not, we'll assume no news is good news, and close this issue in a couple of days.  If so, just let us know, and we'll go from there.

**Comment by ajpowelsnl** [comment]
@Resacd -- please re-open if you're still having issues.

**Comment by sketch34** [comment]
@ajpowelsnl I just encountered this exact same issue trying to compile Kokkos `master` branch with MSVC 19.32.31332.0.

I did a quick experiment and commented the 2nd template declaration. This caused the build to complete without errors.

It looks like the issue may be that the `std::enable_if` logic in the code pasted by @Resacd is resulting in two identical struct definitions.

I'm guessing that 

`std::enable_if<0 < sizeof(&F::shmem_size)>::type>` and
`std::enable_if<0 < sizeof(&F::team_shmem_size)>::type>`

are both evaluating to `true`, resulting in the two identical struct template instantiations. Only the function body is different between them, the signatures are identical.

```
[main] Building folder: kokkos clean
[build] Starting build
[proc] Executing command: C:\Dev\Tools\CMake\bin\cmake.EXE --build c:/Dev/Projects/kokkos/out/build/win-x64 --config Debug --verbose --target clean
[build] Microsoft (R) Build Engine version 17.2.1+52cd2da31 for .NET Framework
[build] Copyright (C) Microsoft Corporation. All rights reserved.
[build] 
[build] Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
[build] Build started 27/07/2022 12:32:26 AM.
[build] Project "C:\Dev\Projects\kokkos\out\build\win-x64\Kokkos.sln" on node 1 (Clean target(s)).
[build] ValidateSolutionConfiguration:
[build]   Building solution configuration "Debug|x64".
[build] ValidateProjects:
[build]   The project "Continuous" is not selected for building in solution configuration "Debug|x64".
[build]   The project "Experimental" is not selected for building in solution configuration "Debug|x64".
[build]   The project "INSTALL" is not selected for building in solution configuration "Debug|x64".
[build]   The project "Nightly" is not selected for building in solution configuration "Debug|x64".
[build]   The project "NightlyMemoryCheck" is not selected for building in solution configuration "Debug|x64".
[build]   The project "RUN_TESTS" is not selected for building in solution configuration "Debug|x64".
[build] Project "C:\Dev\Projects\kokkos\out\build\win-x64\Kokkos.sln" (1) is building "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj.metaproj" (2) on node 1 (Clean target(s)).
[build] Project "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj.metaproj" (2) is building "C:\Dev\Projects\kokkos\out\build\win-x64\ZERO_CHECK.vcxproj" (3) on node 1 (Clean target(s)).
[build] _PrepareForClean:
[build]   Deleting file "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
[build] Done Building Project "C:\Dev\Projects\kokkos\out\build\win-x64\ZERO_CHECK.vcxproj" (Clean target(s)).
[build] Project "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj.metaproj" (2) is building "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj" (4) on node 1 (Clean target(s)).
[build] _PrepareForClean:
[build]   Deleting file "kokkoscore.dir\Debug\kokkoscore.tlog\kokkoscore.lastbuildstate".
[build] Done Building Project "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj" (Clean target(s)).
[build] Done Building Project "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj.metaproj" (Clean target(s)).
[build] Done Building Project "C:\Dev\Projects\kokkos\out\build\win-x64\Kokkos.sln" (Clean target(s)).
[build] 
[build] Build succeeded.
[build]     0 Warning(s)
[build]     0 Error(s)
[build] 
[build] Time Elapsed 00:00:00.18
[build] Build finished with exit code 0
[main] Building folder: kokkos 
[build] Starting build
[proc] Executing command: C:\Dev\Tools\CMake\bin\cmake.EXE --build c:/Dev/Projects/kokkos/out/build/win-x64 --config Debug --verbose --target kokkoscore
[build] Microsoft (R) Build Engine version 17.2.1+52cd2da31 for .NET Framework
[build] Copyright (C) Microsoft Corporation. All rights reserved.
[build] 
[build] Build started 27/07/2022 12:32:26 AM.
[build] Project "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj" on node 1 (default targets).
[build] Project "C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj" (1) is building "C:\Dev\Projects\kokkos\out\build\win-x64\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
[build] InitializeBuildStatus:
[build]   Creating "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
[build] CustomBuild:
[build]   Checking Build System
[build] FinalizeBuildStatus:
[build]   Deleting file "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
[build]   Touching "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
[build] Done Building Project "C:\Dev\Projects\kokkos\out\build\win-x64\ZERO_CHECK.vcxproj" (default targets).
[build] InitializeBuildStatus:
[build]   Creating "kokkoscore.dir\Debug\kokkoscore.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
[build] CustomBuild:
[build]   Building Custom Rule C:/Dev/Projects/kokkos/core/src/CMakeLists.txt
[build] VcpkgTripletSelection:
[build]   Using triplet "x64-windows" from "C:\Dev\Projects\vcpkg\installed\x64-windows\"
[build] ClCompile:
[build]   C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.32.31326\bin\HostX64\x64\CL.exe /c /I"C:\Dev\Projects\kokkos\out\build\win-x64\core\src" /IC:\Dev\Projects\kokkos\core\src /I"C:\Dev\Projects\kokkos\out\build\win-x64" /I"C:\Dev\Projects\vcpkg\installed\x64-windows\include" /Zi /nologo /W1 /WX- /diagnostics:column /Od /Ob0 /D _MBCS /D WIN32 /D _WINDOWS /D KOKKOS_DEPENDENCE /D "CMAKE_INTDIR=\"Debug\"" /Gm- /EHsc /RTC1 /MDd /GS /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"kokkoscore.dir\Debug\\" /Fd"C:\Dev\Projects\kokkos\out\build\win-x64\core\src\Debug\kokkoscore.pdb" /external:W1 /Gd /TP /errorReport:queue C:\Dev\Projects\kokkos\core\src\impl\Kokkos_CPUDiscovery.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Command_Line_Parsing.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Core.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Error.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_ExecPolicy.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_HostBarrier.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_HostSpace.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_HostSpace_deepcopy.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_HostThreadTeam.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_MemoryPool.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_MemorySpace.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_NumericTraits.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Profiling.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Serial.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Serial_Task.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_SharedAlloc.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Spinwait.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_Stacktrace.cpp C:\Dev\Projects\kokkos\core\src\impl\Kokkos_hwloc.cpp C:\Dev\Projects\kokkos\core\src\desul\src\Lock_Array_CUDA.cpp C:\Dev\Projects\kokkos\core\src\desul\src\Lock_Array_HIP.cpp
[build]   Kokkos_CPUDiscovery.cpp
[build]   Kokkos_Command_Line_Parsing.cpp
[build]   Kokkos_Core.cpp
[build] C:\Dev\Projects\kokkos\core\src\impl/Kokkos_FunctorAnalysis.hpp(704,4): error C2953: 'Kokkos::Impl::FunctorAnalysis<PatternInterface,Policy,Functor>::DeduceTeamShmem<F,std::enable_if<0<,void>::type>': class template has already been defined [C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj]
[build] C:\Dev\Projects\kokkos\core\src\impl/Kokkos_FunctorAnalysis.hpp(685): message : see declaration of 'Kokkos::Impl::FunctorAnalysis<PatternInterface,Policy,Functor>::DeduceTeamShmem<F,std::enable_if<0<,void>::type>' [C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj]
[build] C:\Dev\Projects\kokkos\core\src\impl/Kokkos_FunctorAnalysis.hpp(807): message : see reference to class template instantiation 'Kokkos::Impl::FunctorAnalysis<PatternInterface,Policy,Functor>' being compiled [C:\Dev\Projects\kokkos\out\build\win-x64\core\src\kokkoscore.vcxproj]
...
```

**Comment by ajpowelsnl** [comment]
@sketch34 -- many thanks for the head's up on this.  We'll look into it.

**Comment by crtrott** [comment]
Actually, a fix for this is part of #5261 it depends on the exact version of MSVC you have, and now I had it actually pop up in Kokkos tests itself with a specific compiler.
